app doesnt start up after update

hey since the last update to version 1.1.1 the pfa app doesnt start on my pixel 5 anymore.  any way to fix this or secure the notes? please i have a lot of notes

I have the same issue. I updated from v1.0.2 with about 40 notes in it. I have no idea if versions v1.1.x from mid of July were usable. But v1.2.3 to v1.2.5 from beginning of August crash immediately after start.
I did a backup with "Privacy Friendly Backup" (PFB) about nine days ago with v1.2.3. The backup can be exported and it is readable within a text editor (but it is a pain).
I installed v1.2.5 on a second Android device. It is completely useable. But as soon I import above backup on my second device with help of PFB the app crashes here too.
Yesterday I installed F-Droid on my second device to downgrade to an older version of PFN bot all versions smaller than v1.2.3 disappeared at last weekend. But probably this will not help anyhow. Before installing a different version the old one needs to be deinstalled and this will also delete the database. My devices are not rooted so I cannot reach / save the DB files directly.

@Kamuno: Above you asked for a stack trace. Here is mine from my first device with the original database. There I never tried an import. The program complains about database schema. Some columns have a different order and some of "my" fields (e.g. _id) allow NULL.

> 2022-08-15 12:02:31.792 1400-2418/? E/ActivityTrigger: activityStartTrigger: not whiteListedorg.secuso.privacyfriendlynotes/org.secuso.privacyfriendlynotes.ui.SplashActivity/11
2022-08-15 12:02:31.792 1400-2418/? E/ActivityTrigger: activityResumeTrigger: not whiteListedorg.secuso.privacyfriendlynotes/org.secuso.privacyfriendlynotes.ui.SplashActivity/11
2022-08-15 12:02:31.799 1400-2550/? E/ActivityTrigger: activityResumeTrigger: not whiteListedorg.secuso.privacyfriendlynotes/org.secuso.privacyfriendlynotes.ui.SplashActivity/11
2022-08-15 12:02:31.901 1400-3677/? E/ActivityTrigger: activityStartTrigger: not whiteListedorg.secuso.privacyfriendlynotes/org.secuso.privacyfriendlynotes.ui.main.MainActivity/11
2022-08-15 12:02:31.902 1400-3677/? E/ActivityTrigger: activityResumeTrigger: not whiteListedorg.secuso.privacyfriendlynotes/org.secuso.privacyfriendlynotes.ui.main.MainActivity/11
2022-08-15 12:02:31.915 1400-2550/? E/ActivityTrigger: activityResumeTrigger: not whiteListedorg.secuso.privacyfriendlynotes/org.secuso.privacyfriendlynotes.ui.main.MainActivity/11
2022-08-15 12:02:32.017 29608-29636/? E/AndroidRuntime: FATAL EXCEPTION: arch_disk_io_0
    Process: org.secuso.privacyfriendlynotes, PID: 29608
    java.lang.RuntimeException: Exception while computing database live data.
        at androidx.room.RoomTrackingLiveData$1.run(RoomTrackingLiveData.java:92)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.IllegalStateException: Pre-packaged database has an invalid schema: notes(org.secuso.privacyfriendlynotes.room.model.Note).
     Expected:
    TableInfo{name='notes', columns={name=Column{name='name', type='TEXT', affinity='2', notNull=true, primaryKeyPosition=0, defaultValue='null'}, _id=Column{name='_id', type='INTEGER', affinity='3', notNull=true, primaryKeyPosition=1, defaultValue='null'}, in_trash=Column{name='in_trash', type='INTEGER', affinity='3', notNull=true, primaryKeyPosition=0, defaultValue='null'}, type=Column{name='type', type='INTEGER', affinity='3', notNull=true, primaryKeyPosition=0, defaultValue='null'}, category=Column{name='category', type='INTEGER', affinity='3', notNull=true, primaryKeyPosition=0, defaultValue='null'}, content=Column{name='content', type='TEXT', affinity='2', notNull=true, primaryKeyPosition=0, defaultValue='null'}}, foreignKeys=[], indices=[]}
     Found:
    TableInfo{name='notes', columns={in_trash=Column{name='in_trash', type='INTEGER', affinity='3', notNull=true, primaryKeyPosition=0, defaultValue='0'}, name=Column{name='name', type='TEXT', affinity='2', notNull=true, primaryKeyPosition=0, defaultValue='null'}, _id=Column{name='_id', type='INTEGER', affinity='3', notNull=false, primaryKeyPosition=1, defaultValue='null'}, type=Column{name='type', type='INTEGER', affinity='3', notNull=true, primaryKeyPosition=0, defaultValue='null'}, category=Column{name='category', type='INTEGER', affinity='3', notNull=false, primaryKeyPosition=0, defaultValue='null'}, content=Column{name='content', type='TEXT', affinity='2', notNull=true, primaryKeyPosition=0, defaultValue='null'}}, foreignKeys=[], indices=[]}
        at androidx.room.RoomOpenHelper.checkIdentity(RoomOpenHelper.java:163)
        at androidx.room.RoomOpenHelper.onOpen(RoomOpenHelper.java:135)
        at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.onOpen(FrameworkSQLiteOpenHelper.java:201)
        at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:409)
        at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:298)
        at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getWritableSupportDatabase(FrameworkSQLiteOpenHelper.java:151)
        at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper.getWritableDatabase(FrameworkSQLiteOpenHelper.java:112)
        at androidx.room.RoomDatabase.inTransaction(RoomDatabase.java:706)
        at androidx.room.RoomDatabase.assertNotSuspendingTransaction(RoomDatabase.java:483)
        at androidx.room.RoomDatabase.query(RoomDatabase.java:526)
        at androidx.room.util.DBUtil.query(DBUtil.java:86)
        at org.secuso.privacyfriendlynotes.room.dao.NoteDao_Impl$6.call(NoteDao_Impl.java:249)
        at org.secuso.privacyfriendlynotes.room.dao.NoteDao_Impl$6.call(NoteDao_Impl.java:246)
        at androidx.room.RoomTrackingLiveData$1.run(RoomTrackingLiveData.java:90)
        	... 3 more
2022-08-15 12:02:32.043 1400-2110/? E/InputDispatcher: channel '78db74 org.secuso.privacyfriendlynotes/org.secuso.privacyfriendlynotes.ui.main.MainActivity (server)' ~ Channel is unrecoverably broken and will be disposed!

I hope this will help to identify the issue.

--- a/app/src/main/java/org/secuso/privacyfriendlynotes/room/NoteDatabase.java
+++ b/app/src/main/java/org/secuso/privacyfriendlynotes/room/NoteDatabase.java
@@ -50,15 +50,18 @@ public abstract class NoteDatabase extends RoomDatabase {
     public static final int VERSION = 4;
     public static final String DATABASE_NAME = "allthenotes";
     private static NoteDatabase instance;
+
     public abstract NoteDao noteDao();
+
     public abstract CategoryDao categoryDao();
+
     public abstract NotificationDao notificationDao();
 
-    public static synchronized NoteDatabase getInstance(Context context){
+    public static synchronized NoteDatabase getInstance(Context context) {
         return getInstance(context, DATABASE_NAME);
     }
 
-    public static synchronized NoteDatabase getInstance(Context context, String databaseName){
+    public static synchronized NoteDatabase getInstance(Context context, String databaseName) {
         if (instance == null || !DATABASE_NAME.equals(databaseName)) {
             instance = Room.databaseBuilder(context.getApplicationContext(),
                             NoteDatabase.class, databaseName)
@@ -70,8 +73,8 @@ public abstract class NoteDatabase extends RoomDatabase {
         return instance;
     }
 
-    public static synchronized NoteDatabase getInstance(Context context, String databaseName, File file){
-        if (instance == null){
+    public static synchronized NoteDatabase getInstance(Context context, String databaseName, File file) {
+        if (instance == null) {
             instance = Room.databaseBuilder(context.getApplicationContext(),
                             NoteDatabase.class, databaseName)
                     .createFromFile(file)
@@ -100,8 +103,8 @@ public abstract class NoteDatabase extends RoomDatabase {
             // get current schema and check if it needs to be fixed
             String result = "";
             Cursor c = database.query("SELECT sql FROM sqlite_master WHERE type='table' AND name='notes';");
-            if(c != null) {
-                if(c.moveToFirst()) {
+            if (c != null) {
+                if (c.moveToFirst()) {
                     while (!c.isAfterLast()) {
                         result = c.getString(c.getColumnIndexOrThrow("sql"));
                         c.moveToNext();
@@ -110,14 +113,10 @@ public abstract class NoteDatabase extends RoomDatabase {
                 c.close();
             }
 
-            try {
-                String categorySQL = result.split("category")[1].split(",")[0];
+            String categorySQL = result.split("category")[1].split(",")[0];
 
-                if(categorySQL != null && !TextUtils.isEmpty(categorySQL) && !categorySQL.contains("NOT NULL")) {
-                    MIGRATION_1_2.migrate(database);
-                }
-            } catch (NullPointerException e) {
-                // do nothing
+            if (categorySQL != null && categorySQL.toUpperCase().contains("INTEGER") && !categorySQL.toUpperCase().contains("NOT NULL")) {
+                MIGRATION_1_2.migrate(database);
             }
         }
     };
@@ -162,8 +161,8 @@ public abstract class NoteDatabase extends RoomDatabase {
 
             Pair<Integer, String>[] encodedContent = new Pair[0];
             Cursor c = database.query("SELECT * FROM notes WHERE type = 1");
-            if(c != null) {
-                if(c.moveToFirst()) {
+            if (c != null) {
+                if (c.moveToFirst()) {
                     encodedContent = new Pair[c.getCount()];
                     int i = 0;
 
@@ -178,14 +177,14 @@ public abstract class NoteDatabase extends RoomDatabase {
                 c.close();
             }
 
-            for(Pair<Integer,String> note : encodedContent) {
+            for (Pair<Integer, String> note : encodedContent) {
                 ContentValues cv = new ContentValues();
                 cv.put("content", note.second);
                 database.update("notes", 0, cv, "_id = ?", new String[]{Integer.toString(note.first)});
             }
         }
     };
-    static final Migration MIGRATION_1_3 = new Migration(1,3) {
+    static final Migration MIGRATION_1_3 = new Migration(1, 3) {
         @Override
         public void migrate(@NonNull SupportSQLiteDatabase database) {
             MIGRATION_1_2.migrate(database);
@@ -196,15 +195,15 @@ public abstract class NoteDatabase extends RoomDatabase {
         public void migrate(SupportSQLiteDatabase database) {
             Pair<Integer, String>[] encodedContent = new Pair[0];
             Cursor c = database.query("SELECT * FROM notes WHERE type = 1");
-            if(c != null) {
-                if(c.moveToFirst()) {
+            if (c != null) {
+                if (c.moveToFirst()) {
                     encodedContent = new Pair[c.getCount()];
                     int i = 0;
 
                     while (!c.isAfterLast()) {
                         String note = c.getString(c.getColumnIndexOrThrow("content"));
 
-                        if(note.startsWith("<p dir=")) {
+                        if (note.startsWith("<p dir=")) {
                             i++;
                             c.moveToNext();
                             continue;
@@ -219,8 +218,8 @@ public abstract class NoteDatabase extends RoomDatabase {
                 c.close();
             }
 
-            for(Pair<Integer,String> note : encodedContent) {
-                if(note == null) {
+            for (Pair<Integer, String> note : encodedContent) {
+                if (note == null) {
                     continue;
                 }
                 ContentValues cv = new ContentValues();

--- a/app/src/main/java/org/secuso/privacyfriendlynotes/room/NoteDatabase.java
+++ b/app/src/main/java/org/secuso/privacyfriendlynotes/room/NoteDatabase.java
@@ -18,6 +18,7 @@ import android.content.Context;
 import android.database.Cursor;
 import android.text.Html;
 import android.text.SpannedString;
+import android.text.TextUtils;
 
 import androidx.annotation.NonNull;
 import androidx.core.util.Pair;
@@ -25,6 +26,7 @@ import androidx.room.Database;
 import androidx.room.Room;
 import androidx.room.RoomDatabase;
 import androidx.room.migration.Migration;
+import androidx.room.util.StringUtil;
 import androidx.sqlite.db.SupportSQLiteDatabase;
 
 import org.secuso.privacyfriendlynotes.room.dao.CategoryDao;
@@ -35,6 +37,8 @@ import org.secuso.privacyfriendlynotes.room.model.Note;
 import org.secuso.privacyfriendlynotes.room.model.Notification;
 
 import java.io.File;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 /**
  * The database that includes all used information like notes, notifications and categories.
@@ -43,10 +47,10 @@ import java.io.File;
 @Database(
         entities = {Note.class, Category.class, Notification.class},
         version = NoteDatabase.VERSION
-        )
+)
 public abstract class NoteDatabase extends RoomDatabase {
 
-    public static final int VERSION = 3;
+    public static final int VERSION = 4;
     public static final String DATABASE_NAME = "allthenotes";
     private static NoteDatabase instance;
     public abstract NoteDao noteDao();
@@ -60,7 +64,7 @@ public abstract class NoteDatabase extends RoomDatabase {
     public static synchronized NoteDatabase getInstance(Context context, String databaseName){
         if (instance == null || !DATABASE_NAME.equals(databaseName)) {
             instance = Room.databaseBuilder(context.getApplicationContext(),
-                    NoteDatabase.class, databaseName)
+                            NoteDatabase.class, databaseName)
                     .allowMainThreadQueries()
                     .addMigrations(MIGRATIONS)
                     .addCallback(roomCallback)
@@ -72,7 +76,7 @@ public abstract class NoteDatabase extends RoomDatabase {
     public static synchronized NoteDatabase getInstance(Context context, String databaseName, File file){
         if (instance == null){
             instance = Room.databaseBuilder(context.getApplicationContext(),
-                    NoteDatabase.class, databaseName)
+                            NoteDatabase.class, databaseName)
                     .createFromFile(file)
                     .allowMainThreadQueries()
                     .addMigrations(MIGRATIONS)
@@ -86,6 +90,44 @@ public abstract class NoteDatabase extends RoomDatabase {
         @Override
         public void onCreate(@NonNull SupportSQLiteDatabase db) {
             super.onCreate(db);
+        }
+    };
+
+    /**
+     * Provides data migration from database version 3 to 4 which checks for an error in the previous
+     * migration when a backup was imported
+     */
+    static final Migration MIGRATION_3_4 = new Migration(3, 4) {
+        @Override
+        public void migrate(SupportSQLiteDatabase database) {
+            // get current schema and check if it needs to be fixed
+            String result = "";
+            Cursor c = database.query("SELECT sql FROM sqlite_master WHERE type='table' AND name='notes';");
+            if(c != null) {
+                if(c.moveToFirst()) {
+                    while (!c.isAfterLast()) {
+                        result = c.getString(c.getColumnIndexOrThrow("sql"));
+                        c.moveToNext();
+                    }
+                }
+                c.close();
+            }
+
+            try {
+                Matcher m = Pattern.compile("category(.*),", 0).matcher(result);
+
+                String categorySQL = "";
+                if(m.find()) {
+                    categorySQL = m.group(1);
+                }
+
+                if(categorySQL != null && !TextUtils.isEmpty(categorySQL) && !categorySQL.contains("NOT NULL")) {
+                    MIGRATION_1_2.migrate(database);
+                }
+            } catch (NullPointerException e) {
+                // do nothing
+            }
+            System.exit(0);
         }
     };
 
@@ -200,6 +242,7 @@ public abstract class NoteDatabase extends RoomDatabase {
     public static final Migration[] MIGRATIONS = {
             MIGRATION_1_2,
             MIGRATION_1_3,
-            MIGRATION_2_3
+            MIGRATION_2_3,
+            MIGRATION_3_4
     };
 }

--- a/app/src/main/java/org/secuso/privacyfriendlynotes/room/NoteDatabase.java
+++ b/app/src/main/java/org/secuso/privacyfriendlynotes/room/NoteDatabase.java
@@ -26,7 +26,6 @@ import androidx.room.Database;
 import androidx.room.Room;
 import androidx.room.RoomDatabase;
 import androidx.room.migration.Migration;
-import androidx.room.util.StringUtil;
 import androidx.sqlite.db.SupportSQLiteDatabase;
 
 import org.secuso.privacyfriendlynotes.room.dao.CategoryDao;
@@ -37,8 +36,6 @@ import org.secuso.privacyfriendlynotes.room.model.Note;
 import org.secuso.privacyfriendlynotes.room.model.Notification;
 
 import java.io.File;
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
 
 /**
  * The database that includes all used information like notes, notifications and categories.
@@ -114,12 +111,7 @@ public abstract class NoteDatabase extends RoomDatabase {
             }
 
             try {
-                Matcher m = Pattern.compile("category(.*),", 0).matcher(result);
-
-                String categorySQL = "";
-                if(m.find()) {
-                    categorySQL = m.group(1);
-                }
+                String categorySQL = result.split("category")[1].split(",")[0];
 
                 if(categorySQL != null && !TextUtils.isEmpty(categorySQL) && !categorySQL.contains("NOT NULL")) {
                     MIGRATION_1_2.migrate(database);
@@ -127,7 +119,6 @@ public abstract class NoteDatabase extends RoomDatabase {
             } catch (NullPointerException e) {
                 // do nothing
             }
-            System.exit(0);
         }
     };
 

